e2e/rescheduling/rescheduling_test.go

```diff
@@ -0,0 +1,543 @@
+// config that will autorevert on failure
+func TestRescheduling_MaxParallelAutoRevert(t *testing.T) {
+
+	jobID := "test-reschedule-maxp-revert-" + uuid.Generate()[0:8]
+	must.NoError(t, e2eutil.Register(jobID, "./input/rescheduling_maxp_autorevert.nomad"))
+
+	t.Cleanup(func() {
+		e2eutil.StopJob(jobID, "-purge", "-detach")
+		_, err := e2eutil.Command("nomad", "system", "gc")
+		test.NoError(t, err)
+	})
+
+	expected := []string{"running", "running", "running"}
+	must.NoError(t,
+		e2eutil.WaitForAllocStatusExpected(jobID, ns, expected),
+		must.Sprint("should have exactly 3 running allocs"),
+	)
+
+	must.NoError(t,
+		e2eutil.WaitForLastDeploymentStatus(jobID, ns, "successful", nil),
+		must.Sprint("deployment should be successful"))
+
+	// reschedule to make fail
+	job, err := jobspec.ParseFile("./input/rescheduling_maxp_autorevert.nomad")
+	must.NoError(t, err)
+	job.ID = &jobID
+	job.TaskGroups[0].Tasks[0].Config["args"] = []string{"-c", "lol"}
+
+	nc := e2eutil.NomadClient(t)
+	_, _, err = nc.Jobs().Register(job, nil)
+	must.NoError(t, err, must.Sprint("could not e2eutil.Register updated job"))
+
+	must.NoError(t,
+		e2eutil.WaitForAllocStatusComparison(
+			func() ([]string, error) { return e2eutil.AllocStatusesRescheduled(jobID, ns) },
+			func(got []string) bool { return len(got) > 0 }, nil,
+		),
+		must.Sprint("should have new allocs after update"),
+	)
+
+	// wait for the revert
+	expected = []string{"complete", "failed", "running", "running", "running"}
+	must.NoError(t,
+		e2eutil.WaitForAllocStatusComparison(
+			func() ([]string, error) { return e2eutil.AllocStatuses(jobID, ns) },
+			func(got []string) bool {
+				sort.Strings(got)
+				return reflect.DeepEqual(got, expected)
+			}, nil,
+		),
+		must.Sprint("should have one successful, one failed, and 3 reverted allocs"),
+	)
+
+	// at this point the allocs have been checked but we need to wait for the
+	// deployment to be marked complete before we can assert that it's successful
+	// and verify the count of deployments
+	must.NoError(t,
+		e2eutil.WaitForLastDeploymentStatus(jobID, ns, "successful", nil),
+		must.Sprint("most recent deployment should be successful"))
+
+	out, err := e2eutil.Command("nomad", "deployment", "status")
+	must.NoError(t, err, must.Sprint("could not get deployment status"))
+
+	results, err := e2eutil.ParseColumns(out)
+	must.NoError(t, err, must.Sprint("could not parse deployment status"))
+	statuses := map[string]int{}
+	for _, row := range results {
+		if row["Job ID"] == jobID {
+			statuses[row["Status"]]++
+		}
+	}
+
+	must.Eq(t, 1, statuses["failed"],
+		must.Sprintf("expected only 1 failed deployment, got:\n%s", out))
+	must.Eq(t, 2, statuses["successful"],
+		must.Sprintf("expected 2 successful deployments, got:\n%s", out))
+}
+
+// TestRescheduling_ProgressDeadline verifies the progress deadline is reset with
+// each healthy allocation, and that a rescheduled allocation does not.
```

Comment:
again not due to your change, but this test doesn't seem to confirm the latter point - that a rescheduled allocation does not reset the progress deadline.  it only checks that deadline _is_ reset by a healthy alloc.

edit: oh, reading below, `TestRescheduling_ProgressDeadlineFail` covers that case, so it's only this comment that's misleading, the code is fine.